packages/medusa/src/services/order.ts

```diff
@@ -525,9 +546,12 @@ class OrderService extends BaseService {
           )
         }
 
+        const paymentSession = cart.payment_sessions?.find(
+          ({ provider_id }) => provider_id === payment.provider_id
+        )
```

Comment:
question: what is going on here?

At this point, we have validated that the payment has been authorized and we have created the payment, so we are right now double-checking that the payment is good to be attached to the order. There shouldn't be a need to get the details from the payment sessions.

--
Maybe we should rethink some of this, but let's not change the logic right now

Reply:
The getStatus method expect to receive a paymentSession and then will pass the data of that paymentSession to the underlying provider. The provider store the data, that will be used to determine the status, on the paymentSession. Therefore passing the payment here while the data are stored on the session seamed to be a bug.

Wdyt?

Reply:
Maybe we should change `PaymentProviderService.getStatus` to take a less rigid structure as param. Right now the method looks like this;
```js
// packages/medusa/src/services/payment-provider.js - L338

  async getStatus(payment) {
    const provider = this.retrieveProvider(payment.provider_id)
    return provider.getStatus(payment.data)
  }
```

but maybe when we convert this to TypeScript we should do:

```ts
// Paymentish is a terrible name but you get the idea :)
type Paymentish = {
  data: Record<string, unknown>
}

  async getStatus(payment: Paymentish): Promise<PaymentStatus> {
    const provider = this.retrieveProvider(payment.provider_id)
    return provider.getStatus(payment.data)
  }
```

Reply:
But the other method of the payment provider store the data in the payment session data ☺️If you look at the payment migrations. So the payment data will miss some data that are store on the payment session

Reply:
So when the payment provider create the session
```
/**
   * Creates a payment session with the given provider.
   * @param {string} providerId - the id of the provider to create payment with
   * @param {Cart} cart - a cart object used to calculate the amount, etc. from
   * @return {Promise} the payment session
   */
  async createSession(providerId, cart) {
    return this.atomicPhase_(async (manager) => {
      const provider = this.retrieveProvider(providerId)
      const sessionData = await provider.createPayment(cart)

      const sessionRepo = manager.getCustomRepository(
        this.paymentSessionRepository_
      )

      const toCreate = {
        cart_id: cart.id,
        provider_id: providerId,
        data: sessionData,
        status: "pending",
      }

      const created = sessionRepo.create(toCreate)
      const result = await sessionRepo.save(created)

      return result
    })
  }
  ```
  and when it update it, it store the data in the session, which means (i think) that the getStatus can only compute the real status with the payment session data. Am i right?

Reply:
When the payment is created, it is "locked in" so we would not expect the session data to change at that point; but as mentioned earlier - I think it would be a good idea to rethink this flow later. 

So if we stick to the current implementation for now let's take this discussion up later to ensure we think the whole flow through thoroughly. Would that be alright with you?

Reply:
Yes you are right, i ll revert that changes and let’s take this discussion up later about the payment flow indeed 👍

Reply:
done